### PR TITLE
Move pointee_info_at from rustc_codegen_llvm to rustc_target.

### LINKED ISSUE #56166: miri function argument passing should use FnAbi
In the miri engine, when evaluating a function call, it can happen that caller and callee do not agree on the type of an argument. In this case, the argument effectively gets transmuted from the caller type to the callee type. However, this is not legal for all types, and whether it is legal depends on horrible details of the ABI. This logic is implemented for codegen, where it is called `FnType`. miri should probably use that same infrastructure.

The relevant code in miri that would need changing [is here](https://github.com/rust-lang/rust/blob/dae6c9364108847393075b57a17fd7eaf3f37600/src/librustc_mir/interpret/terminator.rs#L184-L227). Currently, we only allow argument type punning for the Rust ABI, and we only allow the valid range of a `Scalar`/`ScalarPair` to change -- effectively, we allow one side to have `&T` while the other side has `*const T`.

For the `FnType` side, I do not know much, but @eddyb offered to mentor someone trying to do this cleanup. :)  He also left the following hints:

> move some code from rustc_codegen_ssa and then compare ArgType pairwise.
> 
> The main issue to moving that stuff around is pointee_info_at (it would have to be moved to rustc::ty::layout, as an additional method in TyLayoutMethods):
> 
> https://github.com/rust-lang/rust/blob/dae6c9364108847393075b57a17fd7eaf3f37600/src/librustc_codegen_llvm/abi.rs#L487
>
> Everything else is mostly relying on rustc_target doing the heavy lifting, already.

Cc @oli-obk

### PR DESCRIPTION
**Pull request by saleemjaffer**
Makes progress towards #56166.

This is a continuation of https://github.com/rust-lang/rust/pull/57150.

@oli-obk Should I close the older PR?

### CODE REVIEW

**Comment by saleemjaffer** [comment]
Creating a new PR as I do not have push rights to @wildarch 's repo

**Comment by oli-obk** [comment]
There are still some points unresolved from the last PR. I think something about the associated types and creating a new trait?On 24 Apr 2019 17:49, Saleem Jaffer <notifications@github.com> wrote:Resolves #56166.
This is a continuation of #60237.
@oli-obk Should I close the older PR?



You can view, comment on, or merge this pull request online at:
  https://github.com/rust-lang/rust/pull/60237

Commit Summary
Move pointee_info_at to TyLayoutMethods.Fix typo in src/librustc/ty/layout.rsReturn instead of collecting to mut result.Remove old pointee_info_at body.Add param_env parameter to pointee_info_at.Make line fit within 100 character limit.impl `pointee_info_at` in TyLayout.resolving conflicts

File Changes

    M
    src/librustc/ty/layout.rs
    (136)
  
    M
    src/librustc_codegen_llvm/abi.rs
    (8)
  
    M
    src/librustc_codegen_llvm/context.rs
    (3)
  
    M
    src/librustc_codegen_llvm/type_of.rs
    (134)
  
    M
    src/librustc_target/abi/mod.rs
    (34)
  

Patch Links:
https://github.com/rust-lang/rust/pull/60237.patchhttps://github.com/rust-lang/rust/pull/60237.diff

**Comment by saleemjaffer** [comment]
@oli-obk Yeah, I have not yet fixed any issues. I just rebased on top of master.

File: src/librustc/ty/layout.rs
Comment (saleemjaffer) [inline src/librustc/ty/layout.rs]: @oli-obk I am trying to add this `to_result `. Having a lot of fallout. First time with traits and lifetimes. So I am not able to progress.

This is the latest error I get:
```
error[E0207]: the type parameter `E` is not constrained by the impl trait, self type, or predicates                                                                         
    --> src/librustc/ty/layout.rs:1614:19
     |
1614 | impl<'a, 'tcx, C, E> TyLayoutMethods<'tcx, C> for Ty<'tcx>
     |                   ^ unconstrained type parameter

error: aborting due to previous error
```

Help please!

Comment (oli-obk) [reply src/librustc/ty/layout.rs]: You can make `E` an associated type instead of a generic parameter.

File: src/librustc/ty/layout.rs
Comment (oli-obk) [inline src/librustc/ty/layout.rs]: Instead of a generic parameter you can use `!` for the associated type here, since we know that `to_result` can only result in an `Ok`

Comment (saleemjaffer) [reply src/librustc/ty/layout.rs]: I tried that earlier. I get this:
```
error: expected one of `>`, `const`, identifier, or lifetime, found `!`118/137: rustc                                                                                       
    --> src/librustc/ty/layout.rs:1501:9
     |
1501 | impl<T, !> MaybeResult<T, !> for T {
     |         ^ expected one of `>`, `const`, identifier, or lifetime here

error: aborting due to previous error

```

Comment (saleemjaffer) [reply src/librustc/ty/layout.rs]:
```
error[E0107]: wrong number of type arguments: expected 1, found 2    ] 118/137: rustc                                                                                       
    --> src/librustc/ty/layout.rs:1501:24
     |
1501 | impl<T> MaybeResult<T, !> for T {
     |                        ^ unexpected type argument

```

Comment (saleemjaffer) [reply src/librustc/ty/layout.rs]: I'll read a bit and try to figure these out.

Comment (oli-obk) [reply src/librustc/ty/layout.rs]: you can just leave both of these out. `!` is a concrete type. Use it for the associated type directly without giving a second generic parameter

**Comment by saleemjaffer** [comment]
@oli-obk This is the fallout:

```
error[E0277]: the trait bound `std::result::Result<rustc_target::abi::TyLayout<'_, &ty::TyS<'_>>, ty::layout::LayoutError<'_>>: ty::layout::MaybeResult<rustc_target::abi::TyLayout<'_, &ty::TyS<'_>>, !>` is not satisfied
    --> src/librustc/ty/layout.rs:1245:30
     |
1245 |                 match layout.field(self, i) {
     |                              ^^^^^ the trait `ty::layout::MaybeResult<rustc_target::abi::TyLayout<'_, &ty::TyS<'_>>, !>` is not implemented for `std::result::Result<rustc_target::abi::TyLayout<'_, &ty::TyS<'_>>, ty::layout::LayoutError<'_>>`
     |
     = help: the following implementations were found:
               <std::result::Result<T, E> as ty::layout::MaybeResult<T, E>>
     = note: required because of the requirements on the impl of `rustc_target::abi::TyLayoutMethods<'_, ty::layout::LayoutCx<'_, ty::context::TyCtxt<'_, 'tcx, 'tcx>>>` for `&ty::TyS<'_>`

error[E0277]: the trait bound `std::result::Result<rustc_target::abi::TyLayout<'_, &ty::TyS<'_>>, ty::layout::LayoutError<'_>>: ty::layout::MaybeResult<rustc_target::abi::TyLayout<'_, &ty::TyS<'_>>, !>` is not satisfied
    --> src/librustc/ty/layout.rs:1312:51
     |
1312 |                                            layout.for_variant(self, i))
     |                                                   ^^^^^^^^^^^ the trait `ty::layout::MaybeResult<rustc_target::abi::TyLayout<'_, &ty::TyS<'_>>, !>` is not implemented for `std::result::Result<rustc_target::abi::TyLayout<'_, &ty::TyS<'_>>, ty::layout::LayoutError<'_>>`
     |
     = help: the following implementations were found:
               <std::result::Result<T, E> as ty::layout::MaybeResult<T, E>>
     = note: required because of the requirements on the impl of `rustc_target::abi::TyLayoutMethods<'_, ty::layout::LayoutCx<'_, ty::context::TyCtxt<'_, 'tcx, 'tcx>>>` for `&ty::TyS<'_>`

error[E0277]: the trait bound `std::result::Result<rustc_target::abi::TyLayout<'_, &ty::TyS<'_>>, ty::layout::LayoutError<'_>>: ty::layout::MaybeResult<rustc_target::abi::TyLayout<'_, &ty::TyS<'_>>, !>` is not satisfied
    --> src/librustc/ty/layout.rs:1998:47
     |
1998 |                 return self.find_niche(layout.field(self, 0)?);
     |                                               ^^^^^ the trait `ty::layout::MaybeResult<rustc_target::abi::TyLayout<'_, &ty::TyS<'_>>, !>` is not implemented for `std::result::Result<rustc_target::abi::TyLayout<'_, &ty::TyS<'_>>, ty::layout::LayoutError<'_>>`
     |
     = help: the following implementations were found:
               <std::result::Result<T, E> as ty::layout::MaybeResult<T, E>>
     = note: required because of the requirements on the impl of `rustc_target::abi::TyLayoutMethods<'_, ty::layout::LayoutCx<'_, ty::context::TyCtxt<'_, 'tcx, 'tcx>>>` for `&ty::TyS<'_>`

error[E0277]: the trait bound `std::result::Result<rustc_target::abi::TyLayout<'_, &ty::TyS<'_>>, ty::layout::LayoutError<'_>>: ty::layout::MaybeResult<rustc_target::abi::TyLayout<'_, &ty::TyS<'_>>, !>` is not satisfied
    --> src/librustc/ty/layout.rs:2006:57
     |
2006 |             if let Some(mut c) = self.find_niche(layout.field(self, i)?)? {
     |                                                         ^^^^^ the trait `ty::layout::MaybeResult<rustc_target::abi::TyLayout<'_, &ty::TyS<'_>>, !>` is not implemented for `std::result::Result<rustc_target::abi::TyLayout<'_, &ty::TyS<'_>>, ty::layout::LayoutError<'_>>`
     |
     = help: the following implementations were found:
               <std::result::Result<T, E> as ty::layout::MaybeResult<T, E>>
     = note: required because of the requirements on the impl of `rustc_target::abi::TyLayoutMethods<'_, ty::layout::LayoutCx<'_, ty::context::TyCtxt<'_, 'tcx, 'tcx>>>` for `&ty::TyS<'_>`

```

File: src/librustc/ty/layout.rs
Comment (oli-obk) [inline src/librustc/ty/layout.rs]: So... apparently doing it this way is blocked on lazy normalization: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e6956340922d4df0087baa36b02b1ce3

Let's go with your original suggestion and add a `from(r: Result<T, Self::item>)` method to `MaybeResult`

File: src/librustc/ty/layout.rs
Comment (saleemjaffer) [inline src/librustc/ty/layout.rs]: @oli-obk This is the issue here:
```
error[E0308]: mismatched types==============================>        ] 118/137: rustc                                                                                       
    --> src/librustc/ty/layout.rs:1633:21
     |
1633 | /                     cx.layout_of(this.ty).to_result().map(|layout| {
1634 | |                         assert_eq!(layout.variants, Variants::Single { index });layout
1635 | |                     })
     | |______________________^ expected enum `std::result::Result`, found struct `rustc_target::abi::TyLayout`
     |
     = note: expected type `std::result::Result<std::result::Result<rustc_target::abi::TyLayout<'tcx, &'tcx ty::TyS<'tcx>>, !>, !>`
                found type `std::result::Result<rustc_target::abi::TyLayout<'_, &ty::TyS<'_>>, <<C as rustc_target::abi::LayoutOf>::TyLayout as ty::layout::MaybeResult<rustc_target::abi::TyLayout<'tcx, &'tcx ty::TyS<'tcx>>>>::Item>`
```

Not sure how to fix this. And I'm bugging you a lot for this issue :(

File: src/librustc/ty/layout.rs
Comment (saleemjaffer) [inline src/librustc/ty/layout.rs]: @oli-obk I get this error here:
```
error[E0308]: mismatched types==============================>        ] 118/137: rustc                                                                                       
    --> src/librustc/ty/layout.rs:1692:32
     |
1656 |     fn field(this: TyLayout<'tcx>, cx: &C, i: usize) -> C::TyLayout {
     |                                                         ----------- expected `<C as rustc_target::abi::LayoutOf>::TyLayout` because of return type
...
1692 |                         return ptr_layout;
     |                                ^^^^^^^^^^ expected associated type, found struct `rustc_target::abi::TyLayout`
     |
     = note: expected type `<C as rustc_target::abi::LayoutOf>::TyLayout`
                found type `rustc_target::abi::TyLayout<'_, &ty::TyS<'_>>`
```

Why is `to_result` changing the type here?

File: src/librustc/ty/layout.rs
Comment (oli-obk) [inline src/librustc/ty/layout.rs]: This should be called `Error`

File: src/librustc/ty/layout.rs
Comment (oli-obk) [inline src/librustc/ty/layout.rs]: You could also write this as `let Ok(x) = x; x`

File: src/librustc/ty/layout.rs
Comment (oli-obk) [inline src/librustc/ty/layout.rs]:
```suggestion
                cx.layout_of(this.ty).to_result().map(|layout| {
```

File: src/librustc/ty/layout.rs
Comment (oli-obk) [reply src/librustc/ty/layout.rs]: The error is sadly not telling you what the actual issue is. What you want here is to convert the result back to a `MaybeResult` after mapping it

```rust
return MaybeResult::from(cx.layout_of(ptr_ty).to_result().map(|mut ptr_layout| {
    ptr_layout.ty = this.ty;
    ptr_layout 
}));
```

Comment (saleemjaffer) [reply src/librustc/ty/layout.rs]: @oli-obk Should this be raised as an issue? :o

Comment (oli-obk) [reply src/librustc/ty/layout.rs]: I think we used to have some more aggressive suggestions around that. If you can create a small example, then a diagnostics issue would be great, yes.

**Comment by saleemjaffer** [comment]
https://github.com/rust-lang/rust/pull/57150#issuecomment-472364882

@oli-obk @eddyb I have added the `is_freeze`. Not sure if this is right as you guys were mentioned about not passing `ParamEnv` separately, but I am still passing `ParamEnv` as a parameter.

File: src/librustc/ty/layout.rs
Comment (oli-obk) [inline src/librustc/ty/layout.rs]: @eddyb what do you think about making `is_freeze` take `TyCtxtAt` and make `HasTyCtxt::tcx` return a `TyCtxtAt`?

Comment (eddyb) [reply src/librustc/ty/layout.rs]: I don't think we should bother, for now.
If you want, you can just make it not a method of `Ty`.

File: src/librustc_target/abi/mod.rs
Comment (eddyb) [inline src/librustc_target/abi/mod.rs]: You can remove this, it's always part of `cx`, because all implementors need it in order to implement `layout_of` anyway!

File: src/librustc_target/abi/mod.rs
Comment (eddyb) [inline src/librustc_target/abi/mod.rs]: Missing space before `->`.

File: src/librustc/ty/layout.rs
Comment (eddyb) [inline src/librustc/ty/layout.rs]: You don't need this because `cx` already contains it, you just need a `HasParamEnv` or similar bound on `C`.
(I need to look at the original PR to see what I suggested there)

File: src/librustc/ty/layout.rs
Comment (eddyb) [inline src/librustc/ty/layout.rs]: Can you move the definition of the trait to just after `trait HasTyCtxt`?

File: src/librustc/ty/layout.rs
Comment (eddyb) [inline src/librustc/ty/layout.rs]: Can you move this `impl` to just after the `HasTyCtxt<'gcx> for LayoutCx` one?

File: src/librustc/ty/layout.rs
Comment (eddyb) [inline src/librustc/ty/layout.rs]: Please keep the formatting from before (you can use `Tools -> Rustfmt` on the playground if you're unusure how to format something - e.g. https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7e17e5eef9cf0076872fec35670b86ea)

File: src/librustc/ty/layout.rs
Comment (eddyb) [inline src/librustc/ty/layout.rs]: Even with this much indentation, rustfmt will keep `return MaybeResult::from(cx.layout_of(ptr_ty).to_result().map(` on one line and just have the closure expression indented by itself, e.g.:
```rust
                    return MaybeResult::from(cx.layout_of(ptr_ty).to_result().map(
                        |mut ptr_layout| {
                            ptr_layout.ty = this.ty;
                            ptr_layout
                        },
                    ));
```

File: src/librustc_target/abi/mod.rs
Comment (eddyb) [inline src/librustc_target/abi/mod.rs]: Huh, was this actually not needed? Only `pointee_info_at`?
Or was my original plan to implement `pointee_info_at` in `rustc_target` instead?
I guess I had forgotten about the special-cases in it? A bit surprising, still, though.

File: src/librustc_codegen_ssa/traits/mod.rs
Comment (eddyb) [inline src/librustc_codegen_ssa/traits/mod.rs]: This shouldn't be needed if `CodegenMethods` includes it.

File: src/librustc/ty/layout.rs
Comment (saleemjaffer) [reply src/librustc/ty/layout.rs]: @eddyb actually rustfmt does this
```rust
return MaybeResult::from(cx.layout_of(ptr_ty).to_result().map(|mut ptr_layout| {
        ptr_layout.ty = this.ty;
        ptr_layout
    }));
```

File: src/librustc_target/abi/mod.rs
Comment (saleemjaffer) [reply src/librustc_target/abi/mod.rs]: @eddyb  You did suggest [this](https://github.com/rust-lang/rust/pull/57150#issuecomment-472364882) first, and that's what i tried to do, but then you said [this](https://github.com/rust-lang/rust/pull/60237#discussion_r280834440)

File: src/librustc/ty/layout.rs
Comment (eddyb) [reply src/librustc/ty/layout.rs]: Only if you have less indentation than you actually do here.

File: src/librustc_target/abi/mod.rs
Comment (eddyb) [reply src/librustc_target/abi/mod.rs]: I was referring to the `param_env` argument, not `is_freeze`.
But I think in your first link, I did forget that you can't implement `pointee_info_at` in `rustc_target` *anyway*.

**Review by eddyb — APPROVED** [review]

**Pull request merged**
